### DISCUSSION TITLE
Add Simplified Chinese Documentation

### DISCUSSION THREAD

**Discussion by Yuerchu** [zauberzeug/nicegui#4025] [Ideas / Feature Requests]
Hi!

I noticed that the documentation currently lacks support for Simplified Chinese (zh-CN). As a Simplified Chinese speaker, I would love to help translate the documentation into Simplified Chinese to make this project more accessible to users and contributors from China and other regions where Simplified Chinese is widely used.

I have already translated some content, such as:
[NiceGUI 文本元素 NiceGUI Text Elements](https://www.yxqi.cn/archives/167)
[NiceGUI ui.label](https://www.yxqi.cn/archives/181)

This is a translation of what was written on my personal blog. Since my blog is based on WordPress, it is not easy to preview the interactive components of NiceGUI.

I hope to upload my translated documentation to the repository in the form of a Pull Request, so that users can choose between Simplified Chinese or English on the official NiceGUI website.

**Comment by Yuerchu** [comment]
Like this (changed by F12 develop mode)
![image](https://github.com/user-attachments/assets/7596d31c-9726-4afe-abc2-b56424fea5f1)

**Reply by python-and-novella** [reply]
Maybe adding i18n api to website is better.

**Reply by rodja** [reply]
How would that be done? Is it a complex change or could you create a pull request to show what you mean, @python-and-fiction ?

**Reply by python-and-novella** [reply]
Not only chinese, any non-english document is needed for others.
So, we can add more pages of document which is translated into other language.
It looks like fastapi's website.
https://fastapi.tiangolo.com/
I can click language switch to visit chinese version website whose src is https://github.com/fastapi/fastapi/tree/master/docs/zh .
Some pages is still english, but someone wants to translate, they can create pr more easily.

**Reply by rodja** [reply]
We are not sure what's the best way to implement this. Just copying the website and letting people start creating pull requests feels not right. Ideally we we would have some translation lookup table which is used in the main documentation to pick the right language. Each table could then be auto-translated with an llm and then fine-tuned with pull requests without needing to touch the page content.

**Reply by python-and-novella** [reply]
I think rodja is right. The best way is automatic translating. 
Maybe chinese user can do this by using llm tools? @Yuerchu

**Reply by Yuerchu** [reply]
> I think rodja is right. The best way is automatic translating. 
> Maybe chinese user can do this by using llm tools? @Yuerchu 

I also recommend using llm, but I personally suggest that, if capable, allowing friends from the open source community to review it to find any potential translation errors. For example, certain proper nouns like "Quasar" might be force translated as "类星体" by translation programs, which is incorrect.

**Comment by Yuerchu** [comment]
I have acquired the domain [nicegui.cn](https://nicegui.cn) and built an official counterpart website using this repository. The site is hosted on servers in Hong Kong, China, optimized for fast access from mainland China. While currently working to localize it into Simplified Chinese, most content remains in English. I'm exploring a more sustainable translation strategy, as directly modifying the source code for translations complicates synchronization with upstream updates of NiceGUI.

我购买了一个域名：[nicegui.cn](https://nicegui.cn) ，并使用此仓库搭建起了一个官方同款的网站。部署的服务器在中国香港，在中国大陆的朋友能很快地访问这个网站。我正在尝试把它翻译成简体中文，虽然目前来看，大部分内容依旧是 English. 但我仍旧在探索一种更好的翻译方案，因为目前我直接修改相关代码来完成翻译，这使得NiceGUI在更新时，此文档的更新也是灾难级的。

![image](https://github.com/user-attachments/assets/3b3b4335-e193-40f1-83c5-f25e1fce644d)

**Reply by evnchn** [reply]
I speak Chinese natively so I can help. 

You can put the source code to the Chinese version of the website to another repo, no need to be put inside NiceGUI repo (let's not overload Falko and Rodja any further)

Then, we'll work towards catching up with the main repo /website path with commits.

**Reply by Yuerchu** [reply]
Thank you for the suggestion, @evnchn! I've forked the repository and completed partial translations. [See details at https://github.com/Yuerchu/nicegui/tree/zh-CN](https://github.com/Yuerchu/nicegui/tree/zh-CN)

十分感谢您的提议，我fork了这个仓库，并翻译了部分内容。[详细请参阅](https://github.com/Yuerchu/nicegui/tree/zh-CN)

**Reply by evnchn** [reply]
Hmm, but then I was thinking, maybe we can make the website support internationalization. This have several advantages over our current split approach: 

1. Support switching between languages smoothly, in case our translation is bad and English original text is more accurate
2. Support adding other languages more easily 

But this will be quite challenging...

**Reply by Yuerchu** [reply]
Yes, I have also considered @python-and-novella's idea of creating i18n support, as this approach could provide solutions not only for Chinese and English users but also allow translators of other languages to contribute. By having the server return different language JSON files during frontend rendering, we could avoid creating multiple branches...
However, the engineering effort would undoubtedly be immense. It requires not only separating bilingual JSON versions but also making extensive modifications to the official website's code... This is far beyond what I can handle alone 😅, but "planting trees for future generations to enjoy the shade" — I'll do as much as I can.
Of course, I admit that creating separate websites for different language versions is the least ideal solution 🫠

**Reply by evnchn** [reply]
My proposal: 

Have a MultilingualString class, when invoke str(), reads the locale set and returns the string of such locale.

Would require minimal changes to the documentation code, which I'll be frank, is probably beyond my abilities to read, let alone edit 😅

**Reply by evnchn** [reply]
https://github.com/evnchn/nicegui-i18n-poc

@Yuerchu let's see if this can work. Provided we call str() before passing into any UI element (f-string does that implicitly) then the MultilingualString class figures out the user language and returns the appropriate string.

**Reply by Yuerchu** [reply]
@evnchn Yes, I agree this approach is also a viable option. However, I think using `ui.run_javascript('navigator.language')` to retrieve the user's language might be better? After detecting the user's language, we could have FastAPI return the corresponding language JSON static files, then hand them over to NiceGUI for rendering.

**Reply by evnchn** [reply]
There is no JSON file. Instead, the documentation is handled via Python and some decorator magic logic that I don't want to understand. 

https://github.com/zauberzeug/nicegui/blob/main/website%2Fdocumentation%2Fcontent%2Faudio_documentation.py

Therefore, I think if we can override the string to be MultilingualString, then we can handle the i18n one step before the documentation display logic kicks in, making life easier for me.

**Reply by evnchn** [reply]
> Have a MultilingualString class, when invoke str(), reads the locale set and returns the string of such locale.

We probably don't even need a class. A function would do. Imagine function `return_correct_string(default, dict_of_translations)`

**Comment by Yuerchu** [comment]
Directly modifying the source code of the NiceGUI repository still seems problematic. Especially when rendering the documentation for NiceGUI methods, it directly reads the docstrings from elements rather than storing them in separate documentation files. This means it can only support a single language. If we forcibly change these docstrings to another language (like Simplified Chinese), then everyone would only see the Simplified Chinese documentation on the official website, and even get only Chinese syntax hints in newer IDEs 😅. Unless we create a separate branch and aggressively modify NiceGUI's source code, I can't find another way to achieve this.

**Reply by evnchn** [reply]
Thought: But then if you talk about "docstring", since it is a string, we are doomed since day1, since (1) each function have 1 docstring, (2) even if there were 2 docstrings, the IDE won't be smart enough to switch between the two. 

It may be temping to just have both English and Chinese, but (1) it gets long and (2) it gets longer with more language support, which we are opening the floodgates to with `MultilingualString`

@Yuerchu can you check other libraries which offer documentation in both Chinese and English, how they handle the docstring issue? 

---

But for this:

```py
@doc.demo('Control the audio element', '''
    This demo shows how to play, pause and seek programmatically.
''')
```

I honestly don't see the issue with:

```py
@doc.demo(MultilingualString("Control the audio element", {"zh": "控制音频元素"}), 
          MultilingualString('''
    This demo shows how to play, pause and seek programmatically.
''', {"zh": "此演示展示如何以编程方式播放、暂停和跳转。"}))
```

Or maybe I'm dumb 😅. What do I know 🤷‍♂️.

**Comment by falkoschindler** [comment]
As you have noticed, NiceGUI's documentation is based on a custom framework that allows to create documentation pages with very little overhead. Take https://github.com/zauberzeug/nicegui/blob/main/website/documentation/content/label_documentation.py for example. There's a decorator `@doc.demo()` that can take a UI element or a function and consumes its docstring. Alternatively, you can specify a title and a description. The decorated function defines an interactive demo. Last but not least, `doc.reference()` auto-generates a reference section.

A disadvantage of this system is that there is no separate content management system other that the Python code itself - docstrings and \*_documentation.py files. To add more languages, we would somehow need to replace docstrings with translations, or replace the whole system altogether. But I think there's so much value in having a low-threshold documentation where you can easily add a documentation page after implementing a new UI element. An i18n solution shouldn't come at a too high cost in this regard.

Which brings me to a rather fundamental question: I know that popular libraries like FastAPI take translation very seriously. And I know that not everyone is speaking English (about 70-80%). Maybe Chinese (20-25%) already covers most of the non-english speaking developers. A next step could be Spanish, Japanese and Russian. But we might not need 20 or so languages.

Let me cite some pros and cons from Claude.ai:

> ### Arguments for minimal manual translation (English + Chinese only):
> - **Quality of automatic translation tools**: Modern translation tools like DeepL and Google Translate have significantly improved for technical content, especially for major languages
> - **Maintenance burden**: Each additional language multiplies the maintenance effort when documentation changes
> - **Developer English proficiency**: Most developers worldwide have working knowledge of English
> - **ROI diminishes**: After Chinese, the benefit per language drops significantly
> 
> ### Arguments for additional languages:
> - **User experience**: Native language documentation provides a better experience, particularly for beginners
> - **Market penetration**: Could help adoption in regions with lower English proficiency
> - **Community building**: Shows respect for diverse developer communities

So when thinking about translation for nicegui.io, we _might_ want to focus on a 2-language solution. In contrast to supporting many languages, we could squeeze two languages into a docstring, for example, and let the `@doc.demo()` decorator pick the right one. This is just a rough idea and far from perfect though.

For two languages we could also further explore the idea of a second repository. What about some Git tooling that displays the diff of individual English pages since the corresponding Chinese page has been updates? I guess we could find a similar solution for docstrings.

**Comment by evnchn** [comment]
@falkoschindler Are you fine with having special tags (like hashtags) in the docstring? 

If so, we can have a tag (Think: `#translation_qwkfvnkslkd`) which indicates to the website, hey if about to display this string, check the translation list id `qwkfvnkslkd` to see if there are other languages version. 

Meanwhile, about #4530, we can also have such tags (Think: `#group_binding`, for those methods related to binding), then we can group the methods in an orderly manner in the website. 

Otherwise, I totally don't see how to work around the fact that NiceGUI uses docstring. A string is a string, and it can only do so much...

**Reply by falkoschindler** [reply]
Tags might be a viable option. We could even hide them using HTML comments:
![Screenshot 2025-04-04 at 09 17 16](https://github.com/user-attachments/assets/bd4b4b6a-3ab8-4d0e-b72e-ee49f051e083)

It's still a lot of overhead, adding such markers to every (even single-line) docstring. And docstrings might be automatically identifiable based on the module and object they are defined in. Of course this isn't the case for other strings like the ones we use in `@doc.demo` decorators.

Anyway, there are ways to mark strings if we need to. But I think this is only one part of the equation. What would we actually do with these string IDs? What kind of content management system would we use to organize translations?

Maybe we could auto-generate IDs based on the file it is define in and the method or class name it documents (in case of docstrings) or decorates (in case of `@doc.demo`). This leaves cases like `doc.text` though or even `doc.ui` (see "Map of NiceGUI"). And it gets worse when trying to translate the whole website including every link, button label, tooltip etc.

**Reply by evnchn** [reply]
Disagree on auto-generating identifiable, because they are likely volatile, and updating the string will cause the auto-generated ID to mutate, unlinking the existing translation. 

Also, in the beginning of the project, I can imagine not every string is translated, and the tag system seems better in that regard. See a tag? Translation is available in the website. If not, use Google translate or something. Then, I am not sure if it is a good idea to hide the tag via HTML comment...

On marking the strings, let me re-state the benefits I have come up with so far: 
* For linking a docstring (a plain old string) to potentially a complex object or a lookup function which can then support translation
* For the many methods in an element, group them by function (binding, styling, etc)

**Reply by Yuerchu** [reply]
I noticed an issue with links generated from the reST documentation of elements. When I force-edit the documents to Chinese, the link target doesn't get added correctly. I initially considered using special tags, but ultimately concluded this wouldn't be a viable solution.

For example, the URL should be `nicegui.cn/documentation/概览_overview` but it becomes `nicegui.cn/documentation#___overview`.

<img alt="Screenshot_2025-04-05-09-36-11-026_com.microsoft.emmx-edit.jpg" src="https://github.com/user-attachments/assets/a684050f-93b6-4997-83e3-0488e36b7832" width="300">

<img alt="Screenshot_2025-04-05-09-34-43-742_com.microsoft.emmx-edit.jpg" src="https://github.com/user-attachments/assets/9eba4d7b-6167-41e6-b9e8-81afe4b56c9c" width="300">

**Reply by evnchn** [reply]
what have you done to do the force-edit?

**Reply by v479038280** [reply]
In fact, the translation effect of AI now is already better than that of most people, and I don't know if there is a need to translate it yet

**Reply by evnchn** [reply]
Not all AIs are so good, and not everyone has access to good AIs. 

The Samsung one is... less than ideal. 

Moreover not all AI models have the context. So perhaps a translation done on our side with a customized prompt is better. 

<img alt="Screenshot_20250405_121602_Samsung Internet.png" src="https://github.com/user-attachments/assets/790fc16c-a7e1-45ba-a151-6cc29f21fdb9" width="300">

**Reply by evnchn** [reply]
Also I'll say: I have to work on a project which needs chinese and english translation, for which they don't allow me to work on the project until June, therefore I'm contributing to NiceGUI's i18n in the meantime, and backporting the changes to my project later. 🧠

**Comment by falkoschindler** [comment]
After thinking about a translation management system for humans, automatic text extraction and AI-based translation:

What if we auto-generate an AI-based translation every time we deploy a new release? We could checkout the repository and ask an AI agent to translate every docstring and every text snippet of the documentation. Additional instructions could give context and steer the agent towards the desired result. It might not be perfect, but requires much less maintenance in the long run.

**Reply by evnchn** [reply]
Concerns: 

- The translation of the same chunk of text must NOT change. 
- The translation of a chunk of text which has gotten a minor addition should ideally NOT change anymore than needed. 
- We need human-in-the-loop. We are NOT shipping translations to prod without at least someone looking at them, or making manual edits. 

My idea so far: 

- Intercept the `doc.demo` functionality for automatic text extraction.
- Generate a JSON of all strings and their machine translations.
- Allow for manual editing of the extracted strings.
- Implement a check for new machine translations:
  - Pair existing strings that haven't changed between translation attempts.
  - Identify the remaining strings with the least edit distance.
  - Include these identified strings in the LLM's context along with existing instructions.
- Leave the demo code alone. 
  - If we have a different set of code for different languages, it may make debugging a broken demo to be way harder.
  - `MultilingualString` can't work for that, since it would show up in the source code, which is ugly and unwanted.

**Reply by Yuerchu** [reply]
> After thinking about a translation management system for humans, automatic text extraction and AI-based translation:
> 
> What if we auto-generate an AI-based translation every time we deploy a new release? We could checkout the repository and ask an AI agent to translate every docstring and every text snippet of the documentation. Additional instructions could give context and steer the agent towards the desired result. It might not be perfect, but requires much less maintenance in the long run.

I maintain a relatively conservative stance towards AI.  
Since AI translation cannot guarantee 100% accuracy, if AI generates incorrect translations, it may lead to unpredictable content, causing developers to write erroneous code.  
- If AI must be used, I believe human review is required;  
- Avoid using AI. Display documentation in the corresponding language if contributors have provided translations, and fall back to English if none are available.

**Reply by evnchn** [reply]
> If AI must be used, I believe human review is required;

Yes. Human-in-the-loop, that's what they call in the lingo. 

> Avoid using AI. Display documentation in the corresponding language if contributors have provided translations, and fall back to English if none are available.

Prefer show AI translation with a disclaimer, than to not use AI out-right. There probably won't be a lot of contributors for translation, so yeah.

**Comment by evnchn** [comment]
Since NiceGUI **does not show up at all** under Chinese language Google search (I used a VPN) whatsoever, I believe i18n is still useful for SEO. 

<img width="1429" height="880" alt="image" src="https://github.com/user-attachments/assets/a19de767-ca1e-480b-989f-ffb7011bcac1" />
